### ISSUE TITLE
Problems with workers not releasing GPU resources.

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#616]
Let's suppose that we define a task that uses one GPU and that task creates a TensorFlow graph on its assigned GPU.

Once the task finishes, TensorFlow presumably will not release the memory that it has reserved on that GPU. We may actually have to kill that worker process in order to free the resources.

We should verify that this actually occurs and potentially add mechanisms for killing the workers to address this problem.

cc @Wapaul1

**Comment by Wapaul1** [comment]
Verified. If two different workers try to create a tensorflow graph and session on the same gpu, the second task will fail with an out of memory error when trying to allocate the weights of the network. However, if the same worker runs the task twice, it will succeed.

The above occurs even when explicitly closing the session.

A possible workaround for now is to simply allocate enough memory so that all workers can allocate a task at the same time on a gpu, (i.e there are 10 workers and each task allocates 10% of the memory.)

**Comment by robertnishihara** [comment]
For the workaround, do you mean within each task telling TensorFlow to only use 10% of the GPU?

Another possible solution is to kill workers after they run GPU tasks.

**Comment by Wapaul1** [comment]
Yes. You can tell tensorflow the fraction using 

```
config = tf.ConfigProto()
config.gpu_options.per_process_gpu_memory_fraction = 0.1
session = tf.Session(config=config, ...)
```

**Comment by mitar** [comment]
Seems this is a problem with [nvidia drivers](https://github.com/tensorflow/tensorflow/issues/1727#issuecomment-285815312)? Try upgrading them?

**Comment by mitar** [comment]
Do you delete session object? Do you use `with tf.Session() as session`?

**Comment by robertnishihara** [comment]
Oh interesting, thanks for the link! I've definitely seen this problem even when closing/deleting the session object. However, I haven't tried upgrading the drivers.

That said, even if you can get it working with TensorFlow, other libraries may fail to release resources or some people may not have the latest drivers, so we may want to come up with a solution for this problem anyway.

**Comment by mitar** [comment]
Shouldn't this maybe then be a parameter to `@ray.remote`? `@ray.remote(max_calls_per_worker=1)` would mean that it should close worker after each call. Many other queuing software has similar things to handle memory leaking: http://docs.celeryproject.org/en/latest/userguide/workers.html#max-tasks-per-child-setting

So how many tasks you want to run per worker before you restart the worker.

**Comment by robertnishihara** [comment]
Oh nice, that might be the cleanest way of doing this!

I suppose if the problem with GPU resources not being released is very common, we could also set `max_calls_per_worker=1` by default for GPU tasks.

**Comment by robertnishihara** [comment]
This is addressed by #660.

**Comment by neuronflow** [comment]
this is still mentioned in the docs? is it still true?
https://ray.readthedocs.io/en/stable/using-ray-with-gpus.html#troubleshooting
